src/Elastic.Apm/DiagnosticListeners/HttpDiagnosticListenerImplBase.cs

```diff
@@ -0,0 +1,218 @@
+		/// </summary>
+		internal readonly ConcurrentDictionary<TRequest, Span> ProcessingRequests = new ConcurrentDictionary<TRequest, Span>();
+
+		protected HttpDiagnosticListenerImplBase(IApmAgent components)
+		{
+			_logger = components.Logger?.Scoped("HttpDiagnosticListenerImplBase");
+			_configurationReader = components.ConfigurationReader;
+		}
+
+		protected abstract string RequestGetMethod(TRequest request);
+
+		protected abstract Uri RequestGetUri(TRequest request);
+
+		protected abstract void RequestHeadersAdd(TRequest request, string headerName, string headerValue);
+
+		protected abstract bool RequestHeadersContain(TRequest request, string headerName);
+
+		protected abstract int ResponseGetStatusCode(TResponse response);
+
+		public abstract string Name { get; }
+
+		internal abstract string ExceptionEventKey { get; }
+		internal abstract string StartEventKey { get; }
+		internal abstract string StopEventKey { get; }
+
+		public void OnCompleted() { }
+
+		public void OnError(Exception error) => _logger.Error()?.LogExceptionWithCaller(error, nameof(OnError));
+
+		public void OnNext(KeyValuePair<string, object> kv)
+		{
+			_logger.Trace()?.Log("Called with key: `{DiagnosticEventKey}', value: `{DiagnosticEventValue}'", kv.Key, kv.Value);
+
+			if (string.IsNullOrEmpty(kv.Key))
+			{
+				_logger.Trace()?.Log($"Key is {(kv.Key == null ? "null" : "an empty string")} - exiting");
+				return;
+			}
+
+			if (kv.Value == null)
+			{
+				_logger.Trace()?.Log("Value is null - exiting");
+				return;
+			}
+
+			var requestObject = kv.Value.GetType().GetTypeInfo().GetDeclaredProperty(EventRequestPropertyName)?.GetValue(kv.Value);
+			if (requestObject == null)
+			{
+				_logger.Trace()?.Log("Event's {EventRequestPropertyName} property is null - exiting", EventRequestPropertyName);
+				return;
+			}
+
+			var request = requestObject as TRequest;
+			if (request == null)
+			{
+				_logger.Trace()
+					?.Log("Actual type of object ({EventRequestPropertyActualType}) in event's {EventRequestPropertyName} property " +
+						"doesn't match the expected type ({EventRequestPropertyExpectedType}) - exiting",
+						requestObject.GetType().FullName, EventRequestPropertyName, typeof(TRequest).FullName);
+				return;
+			}
+
+			var requestUrl = RequestGetUri(request);
+			if (requestUrl == null)
+			{
+				_logger.Trace()?.Log("Request URL is null - exiting", EventRequestPropertyName);
+				return;
+			}
+
+			if (IsRequestFilteredOut(requestUrl))
+			{
+				_logger.Trace()?.Log("Request URL ({RequestUrl}) is filtered out - exiting", requestUrl);
+				return;
+			}
+
+			if (kv.Key.Equals(StartEventKey))
+				ProcessStartEvent(kv.Value, request, requestUrl);
+			else if (kv.Key.Equals(StopEventKey))
+				ProcessStopEvent(kv.Value, request, requestUrl);
+			else if (kv.Key.Equals(ExceptionEventKey))
+				ProcessExceptionEvent(kv.Value, request, requestUrl);
+			else
+				_logger.Error()?.Log("Unrecognized key `{DiagnosticEventKey}'", kv.Key);
```

Comment:
Opened https://github.com/elastic/apm-agent-dotnet/pull/224, I think this should not be an error.